File: Classes/CurrencyManager.m
// -*-  Mode:ObjC; c-basic-offset:4; tab-width:8; indent-tabs-mode:nil -*-
/*
 * CashFlow for iOS
 * Copyright (C) 2008-2012, Takuya Murakami, All rights reserved.
 * For conditions of distribution and use, see LICENSE file.
 */

#import "CurrencyManager.h"

@implementation CurrencyManager
{
    NSNumberFormatter *_numberFormatter;
}

/**
 * CurrencyManager のインスタンスを返す
 */
+ (CurrencyManager *)instance
{
    static CurrencyManager *theInstance = nil;
    if (theInstance == nil) {
        theInstance = [CurrencyManager new];
    }
    return theInstance;
}

- (id)init
{
    self = [super init];

    NSNumberFormatter *nf;

    nf = [[NSNumberFormatter alloc] init];
    [nf setNumberStyle:NSNumberFormatterCurrencyStyle];
    [nf setLocale:[NSLocale currentLocale]];
    _numberFormatter = nf;

    self.currencies =
        @[@"AED",
         @"AUD",
         @"BHD",
         @"BND",
         @"BRL",
         @"CAD", 
         @"CHF",
         @"CLP",
         @"CNY",
         @"CYP",
         @"CZK",
         @"DKK",
         @"EUR",
         @"GBP",
         @"HKD",
         @"HUF",
         @"IDR",
         @"ILS",
         @"INR",
         @"ISK",
         @"JPY",
         @"KRW",
         @"KWD",
         @"KZT",
         @"LKR",
         @"MTL",
         @"MUR",
         @"MXN",
         @"MYR",
         @"NOK",
         @"NPR",
         @"NZD",
         @"OMR",
         @"PKR",
         @"QAR",
         @"RUB",
         @"SAR",
         @"SEK",
         @"SGD",
         @"SKK",
         @"THB",
         @"TWD",
         @"USD",
         @"ZAR"];

    self.baseCurrency = [[NSUserDefaults standardUserDefaults] objectForKey:@"BaseCurrency"];

    return self;
}

/**
 * システムデフォルトの通貨コードを返す
 */
+ (NSString *)systemCurrency
{
    NSNumberFormatter *nf = [NSNumberFormatter new];
    [nf setNumberStyle:NSNumberFormatterCurrencyStyle];
    return [nf currencyCode];
}

/**
 * ベース通貨コードを設定する
 */
- (void)setBaseCurrency:(NSString *)currency
{
    if (_baseCurrency != currency) {
        _baseCurrency = currency;
        
        if (currency == nil) {
            currency = [CurrencyManager systemCurrency];
        }
        [_numberFormatter setCurrencyCode:currency];
        
        [[NSUserDefaults standardUserDefaults] setObject:_baseCurrency forKey:@"BaseCurrency"];
    }
}

/**
 * 通貨を文字列にフォーマットする
 */
+ (NSString *)formatCurrency:(double)value
{
    return [[CurrencyManager instance] _formatCurrency:value];
}

- (NSString *)_formatCurrency:(double)value
{
    NSNumber *n = @(value);
    return [_numberFormatter stringFromNumber:n];
}

@end
        

File: Classes/CurrencyManager.swift
//
//  CurrencyManager.swift
//

import Foundation
import UIKit

class CurrencyManager: NSObject {
    let currencies = NSLocale.ISOCurrencyCodes() as [String];
    /*
    ["AED",
    "AUD",
    "BHD",
    "BND",
    "BRL",
    "CAD",
    "CHF",
    "CLP",
    "CNY",
    "CYP",
    "CZK",
    "DKK",
    "EUR",
    "GBP",
    "HKD",
    "HUF",
    "IDR",
    "ILS",
    "INR",
    "ISK",
    "JPY",
    "KRW",
    "KWD",
    "KZT",
    "LKR",
    "MTL",
    "MUR",
    "MXN",
    "MYR",
    "NOK",
    "NPR",
    "NZD",
    "OMR",
    "PKR",
    "QAR",
    "RUB",
    "SAR",
    "SEK",
    "SGD",
    "SKK",
    "THB",
    "TWD",
    "USD",
    "ZAR"]
*/
    
    private let kBaseCurrency = "BaseCurrency"
    
    private let _numberFormatter = NSNumberFormatter()
    
    class var instance: CurrencyManager {
        struct Static {
            static let instance : CurrencyManager = CurrencyManager()
        }
        return Static.instance
    }
    
    class func getInstance() -> CurrencyManager {
        return CurrencyManager.instance
    }
    
    private override init() {
        super.init()
        
        _numberFormatter.numberStyle = .CurrencyStyle
        _numberFormatter.locale = NSLocale.currentLocale()
        
        _baseCurrency = NSUserDefaults.standardUserDefaults().objectForKey(kBaseCurrency) as NSString?

        // TEST : get currency code list
        /*
        var codes = NSLocale.ISOCurrencyCodes() as [String];
        for code in codes {
            NSLog(code);
        }
        */
    }
    
    /**
     * システムデフォルトの通貨コードを返す
     */
    class func systemCurrency() -> String {
        var nf = NSNumberFormatter()
        nf.numberStyle = .CurrencyStyle
        return nf.currencyCode
    }
    
    /**
     * ベース通貨コード
     */
    var baseCurrency: String? {
        get {
            return _baseCurrency
        }
        set {
            if _baseCurrency != newValue {
                _baseCurrency = newValue
            
                if (_baseCurrency != nil){
                    _numberFormatter.currencyCode = _baseCurrency!
                } else {
                    _numberFormatter.currencyCode = CurrencyManager.systemCurrency()
                }
            
                NSUserDefaults.standardUserDefaults().setObject(_baseCurrency, forKey: kBaseCurrency)
            }
        }
    }
    
    private var _baseCurrency: NSString?
    
    class func formatCurrency(value: Double) -> String {
        return CurrencyManager.instance._formatCurrency(value)
    }
    
    private func _formatCurrency(value: Double) -> String {
        return _numberFormatter.stringFromNumber(value)!
    }
}
